### PR TITLE
Add middleware abstraction

### PR DESCRIPTION
**Pull request by BjoernAtBosch**
## Describe your changes

Add first version of middleware abstraction to get rid of C++ specific configuration (e.g. environment variables)

## Issue ticket number and link

tbd

## Checklist - Manual tasks

* [x] Examples are executing successfully
* [x] Created/updated unit tests. Code Coverage percentage on new code shall be >= 80%.
* [ ] Created/updated integration tests.
* [x] Devcontainer can be opened successfully
* [x] Devcontainer can be opened successfully behind a corporate proxy
* [x] Devcontainer can be re-built successfully
* [x] Extended the documentation (e.g. README.md, CONTRIBUTING.md, Velocitas)

### CODE REVIEW

File: sdk/include/sdk/Config.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+#ifndef VEHICLE_APP_SDK_CONFIG_H
+#define VEHICLE_APP_SDK_CONFIG_H
+
+#include "middleware/Middleware.h"
+
+namespace velocitas {
+
+class Config {
```
Comment (kse3hi) [inline sdk/include/sdk/Config.h:24]: It looks like factory or provider, but not as a config :)

File: sdk/src/sdk/middleware/DaprMiddleware.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) 2022-2023 Robert Bosch GmbH
```
Comment (kse3hi) [inline sdk/src/sdk/middleware/DaprMiddleware.cpp:2]: It is the new created file. Shall be only 2023

File: sdk/src/sdk/middleware/DaprMiddleware.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+static constexpr char const* ENV_DAPR_HTTP_PORT = "DAPR_HTTP_PORT";
+
+static constexpr char const* DAPR_APP_ID_KEY = "dapr-app-id";
+
+namespace dapr {
+
+void waitForSidecar() {
+    auto sidecarHttpPort = getEnvVar(ENV_DAPR_HTTP_PORT);
+    if (sidecarHttpPort.empty()) {
+        logger().info("dapr: env {} not set. Continuing without sidecar health check ...",
```
Comment (kse3hi) [inline sdk/src/sdk/middleware/DaprMiddleware.cpp:43]: it shalle be error or at least warning?

File: sdk/src/sdk/middleware/DaprMiddleware.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                constexpr auto STATUS_NO_CONTENT{204};
+                logger().info("dapr: Requesting side car health endpoint...");
+
+                const auto response = cpr::Get(
+                    cpr::Url(fmt::format("http://localhost:{}/v1.0/healthz", sidecarHttpPort)));
+                success = response.status_code == STATUS_NO_CONTENT;
+                logger().debug("dapr: Health endpoint returned status code: {}, {}",
+                               response.status_code, response.text);
+            } catch (const std::exception& e) {
+                logger().debug("dapr: Exception occurred during health endpoint: {}", e.what());
```
Comment (kse3hi) [inline sdk/src/sdk/middleware/DaprMiddleware.cpp:59]: Shall be error or warning?

**Review by kse3hi — COMMENTED** [review]
General question: who is responsible for ENV_VARIABLE definition? 
What about launch.json and tasks.json, is any modification is required?
Shall SDK get values not from ENV_VARIABLES directly, but by getting them out of the application context?

File: sdk/include/sdk/Config.h (hunk 1 above)
Comment (BjoernAtBosch) [reply sdk/include/sdk/Config.h:24]: Yeah, not sure about this. Maybe we should remove it again. It is also there in Python, but I must admit that I didn't spent so much thinking about this.

File: sdk/src/sdk/middleware/DaprMiddleware.cpp (hunk 2 above)
Comment (BjoernAtBosch) [reply sdk/src/sdk/middleware/DaprMiddleware.cpp:2]: I put 2022 here also as I moved the `waitForSidecar` here, which is from 2022

File: sdk/src/sdk/middleware/DaprMiddleware.cpp (hunk 3 above)
Comment (BjoernAtBosch) [reply sdk/src/sdk/middleware/DaprMiddleware.cpp:44]: Yes, could be warning

**Comment by BjoernAtBosch** [comment]
Answer to Serhii`s (kse3hi) general questions:
* Who is responsible for ENV_VARIABLE definition? 
  -> The definer of the runtime.
* What about launch.json and tasks.json, is any modification is required?
  -> No modification required
* Shall SDK get values not from ENV_VARIABLES directly, but by getting them out of the application context?
  -> You mean the app context should parse the env vars? That would mean, that every app has to take care for it and need to handle middleware specifics.
  -> We do it "same" (from user perspective) way in Python SDK.

File: sdk/tests/unit/DaprMiddleware_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+TEST_F(Test_DaprMiddleware, getServiceLocation_envDaprGrpcPortSet_contentOfEnvVar) {
+    ::setenv("DAPR_GRPC_PORT", "12345", /*overwrite=*/true);
+    auto serviceLocation = getCut().getServiceLocation("SomeService");
+    EXPECT_EQ("localhost:12345", serviceLocation);
+}
+
+TEST_F(Test_DaprMiddleware, getServiceLocation_envDaprGrpcPortNotSet_throwsRuntimeError) {
+    ::unsetenv("DAPR_GRPC_PORT");
```
Comment (doosuu) [inline sdk/tests/unit/DaprMiddleware_tests.cpp:51]: Let's use the fixture's `TearDown` method to unset all potentially set env vars.

File: sdk/tests/unit/NativeMiddleware_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    EXPECT_NE(nullptr, pubSubClient.get());
+}
+
+TEST_F(Test_NativeMiddleware, getServiceLocation_envVarNotSet_emptyString) {
+    auto serviceLocation = getCut().getServiceLocation("UnknownService");
+    EXPECT_EQ("", serviceLocation);
+}
+
+TEST_F(Test_NativeMiddleware, getServiceLocation_envVarSet_contentOfEnvVar) {
+    ::setenv("SDV_SOMESERVICE_ADDRESS", "some-service-address", /*overwrite=*/true);
```
Comment (doosuu) [inline sdk/tests/unit/NativeMiddleware_tests.cpp:50]: Let's use the fixture's `TearDown` method to unset all potentially set env vars.

**Review by doosuu — CHANGES_REQUESTED** [review]

File: sdk/src/sdk/middleware/DaprMiddleware.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+static constexpr char const* ENV_DAPR_HTTP_PORT = "DAPR_HTTP_PORT";
+
+static constexpr char const* DAPR_APP_ID_KEY = "dapr-app-id";
+
+namespace dapr {
+
+void waitForSidecar() {
+    auto sidecarHttpPort = getEnvVar(ENV_DAPR_HTTP_PORT);
+    if (sidecarHttpPort.empty()) {
+        logger().warn("dapr: env {} not set. Continuing without sidecar health check ...",
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/middleware/DaprMiddleware.cpp:44]: I guess, yes: Not all of our (incl. VAL) components are doing proper health checking...

File: sdk/src/sdk/middleware/DaprMiddleware.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+#include <thread>
+
+namespace velocitas {
+
+static constexpr char const* ENV_DAPR_GRPC_PORT = "DAPR_GRPC_PORT";
+static constexpr char const* ENV_DAPR_HTTP_PORT = "DAPR_HTTP_PORT";
+
+static constexpr char const* DAPR_APP_ID_KEY = "dapr-app-id";
+
+namespace dapr {
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/middleware/DaprMiddleware.cpp:39]: `waitForSidecar` is part of the public interface. Therefore kept for backwards compatability

File: sdk/src/sdk/middleware/DaprMiddleware.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/middleware/DaprMiddleware.cpp]: As discussed, skipping health check might make sense for (pure) services, but not for app (that typically connect to services). Those have to wait till sidecar is ready.

File: sdk/src/sdk/Model.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "sdk/Model.h"
+#include "sdk/Config.h"
+
+namespace velocitas {
+
+std::string Service::getLocation() const {
+    return Config::getMiddleware().getServiceLocation(getName());
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/Model.cpp]: Would prefer refactoring this in a separate issue (like next finding)

File: sdk/src/sdk/VehicleApp.cpp [hunk 10]
```diff
@@ -40,12 +40,11 @@ VehicleApp::VehicleApp(std::shared_ptr<IVehicleDataBrokerClient> vdbClient,
 }
 
 void VehicleApp::run() {
-    m_pubSubClient->connect();
-
     logger().info("Running App...");
+    Config::getMiddleware().start();
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/VehicleApp.cpp]: Prefer separate issue/improvement for this.

File: sdk/tests/unit/Middleware_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    auto middleware = Middleware::instantiate();
+    EXPECT_EQ(DaprMiddleware::TYPE_ID, middleware->getTypeId());
+}
+
+TEST_F(Test_Middleware, instantiate_envVarSetToDapr_typeIdIsDapr) {
+    setEnvVar(Middleware::getTypeDefiningEnvVarName(), DaprMiddleware::TYPE_ID);
+    EXPECT_EQ(DaprMiddleware::TYPE_ID, Middleware::instantiate()->getTypeId());
+}
+
+TEST_F(Test_Middleware, instantiate_envVarSetToNative_TypeIdIsNative) {
```
Comment (doosuu) [inline sdk/tests/unit/Middleware_tests.cpp:49]:
```suggestion
TEST_F(Test_Middleware, instantiate_envVarSetToNative_typeIdIsNative) {
```

File: sdk/tests/unit/TestBaseUsingEnvVars.cpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ */
+
+#include "TestBaseUsingEnvVars.h"
+
+#include <cstdlib>
+#include <string>
+
+namespace velocitas {
+
+void TestUsingEnvVars::checkToStoreVariableState(const std::string& varName) {
```
Comment (doosuu) [inline sdk/tests/unit/TestBaseUsingEnvVars.cpp:24]: A rather complex method name. Can we think of something better?

E.g.

```suggestion
void TestUsingEnvVars::preserveEnvVarState(const std::string& varName) {
```

File: sdk/tests/unit/NativeMiddleware_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+TEST_F(Test_NativeMiddleware, createPubSubClient__validPointer) {
+    std::shared_ptr<IPubSubClient> pubSubClient;
+    EXPECT_NO_THROW(pubSubClient = getCut().createPubSubClient("My Test Id"));
+    EXPECT_NE(nullptr, pubSubClient.get());
+}
+
+TEST_F(Test_NativeMiddleware, getServiceLocation_envVarNotSet_emptyString) {
+    auto serviceLocation = getCut().getServiceLocation("UnknownService");
```
Comment (doosuu) [inline sdk/tests/unit/NativeMiddleware_tests.cpp:44]: Should this not throw an exception when there is no such service variable?

File: sdk/tests/unit/DaprMiddleware_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+TEST_F(Test_DaprMiddleware, getServiceLocation_envDaprGrpcPortNotSet_throwsRuntimeError) {
+    unsetEnvVar("DAPR_GRPC_PORT");
+    EXPECT_THROW(getCut().getServiceLocation("UnknownService"), std::runtime_error);
+}
+
+TEST_F(Test_DaprMiddleware, getMetadata_envVarWithApppIdSet_metadataWithContentOfEnvVar) {
+    setEnvVar("SOMESERVICE_DAPR_APP_ID", "some-explicit-app-id");
+    Middleware::Metadata metadata         = getCut().getMetadata("SomeService");
+    Middleware::Metadata expectedMetadata = {{"dapr-app-id", "some-explicit-app-id"}};
```
Comment (doosuu) [inline sdk/tests/unit/DaprMiddleware_tests.cpp:57]: Make expectation const

File: sdk/tests/unit/DaprMiddleware_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    EXPECT_EQ(expectedMetadata, metadata);
+}
+
+TEST_F(Test_DaprMiddleware, getMetadata_envVarWithApppIdNotSet_metadataWithDefaultAppId) {
+    unsetEnvVar("SOMESERVICE_DAPR_APP_ID");
+    Middleware::Metadata metadata         = getCut().getMetadata("SomeService");
+    Middleware::Metadata expectedMetadata = {{"dapr-app-id", "someservice"}};
```
Comment (doosuu) [inline sdk/tests/unit/DaprMiddleware_tests.cpp:64]: Make expectation const

File: sdk/include/sdk/middleware/Middleware.h [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+     * @brief Returns a reference to a singelton instance of a concrete middleware class
+     *
+     * @return Middleware&
+     */
+    static Middleware& getInstance() {
+        static std::unique_ptr<Middleware> singleton = instantiate();
+        return *singleton;
+    }
+
+    static std::string getTypeDefiningEnvVarName();
```
Comment (doosuu) [inline sdk/include/sdk/middleware/Middleware.h:43]: Any reason why this is a static method rathern than a constant on header level?

**Review by doosuu — CHANGES_REQUESTED** [review]

File: sdk/tests/unit/Middleware_tests.cpp [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+using namespace velocitas;
+
+class Test_Middleware : public TestUsingEnvVars {};
+
+TEST_F(Test_Middleware, getTypeDefiningEnvVarName__noneEmptyString) {
+    EXPECT_FALSE(Middleware::getTypeDefiningEnvVarName().empty());
+}
+
+TEST_F(Test_Middleware, getInstantance_envVarGloballySetToNative_TypeIdIsNative) {
```
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/tests/unit/Middleware_tests.cpp:33]: Typo fixed, "defaultsToNative" isn't correct in my opinion: It's explicitly set, no default (I think thats true for the next test "instantiate_envVarNotSet_...").

File: sdk/include/sdk/middleware/Middleware.h (hunk 16 above)
Comment (BjoernAtBosch) [reply sdk/include/sdk/middleware/Middleware.h:43]: No clue, why I did it as a function. The definition I prefer to have in the cpp-file.

File: sdk/src/sdk/Model.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/Model.cpp]: Issue created: #66

File: sdk/src/sdk/VehicleApp.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (BjoernAtBosch) [reply sdk/src/sdk/VehicleApp.cpp]: Issue create: #66

**Review by doosuu — APPROVED** [review]
LGTM

**Pull request merged**